### PR TITLE
stream processing base for the producer

### PR DESCRIPTION
**Pull request by rmothilal**
Producer and LIME parser for Stream Processing API

### CODE REVIEW

File: src/stream/producer.js [hunk 1]
```diff
@@ -31,12 +31,296 @@
...
+}
+const
+ENUMS = {
+  EVENTS,
+  METHOD,
+  STATUS
+}
+
+module.exports = ENUMS
+
```
Comment (mdebarros) [inline src/stream/producer.js:112]: missing jsdoc for the constructor and the class?

File: src/stream/producer.js [hunk 2]
```diff
@@ -31,12 +31,296 @@
...
+  prepared: 'prepared',
+  fulfill: 'fulfill',
+  reject: 'reject',
+  position: 'position',
+  notification: 'notification',
+  failed: 'failed',
+  duplicate: 'duplicate'
+}
+
+const METHOD = {
```
Comment (mdebarros) [inline src/stream/producer.js]: add jsdoc

File: src/stream/producer.js [hunk 3]
```diff
@@ -31,12 +31,296 @@
...
 
-// TODO:  RDKAFKA Producer code goes here
+const EventEmitter = require('events')
+const Logger = require('../logger')
+const Kafka = require('node-rdkafka')
+const LimeParser = require('./protocol')
 
-class Producer {
-  constructor () {
+const EVENTS = {
```
Comment (mdebarros) [inline src/stream/producer.js]: add jsdoc

File: src/stream/producer.js [hunk 4]
```diff
@@ -31,12 +31,296 @@
...
+}
+
+const METHOD = {
+  get: 'get',
+  post: 'post',
+  put: 'put',
+  del: 'delete'
+}
+
+const STATUS = {
```
Comment (mdebarros) [inline src/stream/producer.js]: add jsdoc

File: src/stream/producer.js [hunk 5]
```diff
@@ -31,12 +31,296 @@
...
+  METHOD,
+  STATUS
+}
+
+module.exports = ENUMS
+
+class Producer extends EventEmitter {
+  constructor (options = {requiredAcks: -1, partitionCount: 1}, defaultPartitionCount = 1) {
+    super()
+    var config = {
```
Comment (mdebarros) [inline src/stream/producer.js]: this needs to be moved into the constructor

File: src/stream/producer.js [hunk 6]
```diff
@@ -31,12 +31,296 @@
...
+ENUMS = {
+  EVENTS,
+  METHOD,
+  STATUS
+}
+
+module.exports = ENUMS
+
+class Producer extends EventEmitter {
+  constructor (options = {requiredAcks: -1, partitionCount: 1}, defaultPartitionCount = 1) {
```
Comment (mdebarros) [inline src/stream/producer.js]: Whats the difference between partitionCount and defaultPartitionCount?

File: src/stream/producer.js [hunk 7]
```diff
@@ -31,12 +31,296 @@
...
+    if (!config) {
+      throw new Error('missing a config object')
+    }
+    let {logger} = config
+    logger.silly('Producer::constructor() - start')
+    this._config = config
+    this._status = {}
+    this._status.runningInProduceMode = false
+    this._status.runningInProduceBatchMode = false
+    this._producerPollIntv = null
```
Comment (mdebarros) [inline src/stream/producer.js:193]: not used, pls remove

**Comment by rmothilal** [comment]
Please review and notify me on changes required

File: src/stream/producer.js [hunk 8]
```diff
@@ -31,12 +31,384 @@
...
+ * @param {object} options - Key value pairs for mapping to the configuration
+ * @param {object} config - Key value pairs for the configuration of the Producer with the following:
+ * rdkafkaConf - specific rdkafka configurations [Refer to configuration doc]{@link https://github.com/edenhill/librdkafka/blob/0.11.1.x/CONFIGURATION.md}
+ * topicConf - topic configuration
+ * logger - logger object that supports debug(), info(), verbose(), error() & silly()
+ * @extends EventEmitter
+ * @constructor
+ */
+class Producer extends EventEmitter {
+  constructor (options = {requiredAcks: -1, partitionCount: 1, pollIntervalMs: 100}, config = {
```
Comment (mdebarros) [inline src/stream/producer.js]: This does not conform to the standard config approach for the consumer

File: src/stream/producer.js [hunk 9]
```diff
@@ -31,12 +31,384 @@
...
+      'socket.keepalive.enable': options['socket.keepalive.enable'] || true,
+      'queue.buffering.max.messages': options['queue.buffering.max.messages'] || 10,
+      'queue.buffering.max.ms': options['queue.buffering.max.ms'] || 50,
+      'batch.num.messages': options['batch.num.messages'] || 10000,
+      'api.version.request': true,
+      'dr_cb': true
+    },
+    topicConf: {
+      // 0=Broker does not send any response/ack to client, 1=Only the leader broker will need to ack the message, -1 or all=broker will block until message is committed by all in sync replicas (ISRs) or broker's min.insync.replicas setting before sending response.
+      'request.required.acks': options.requiredAcks || 1
```
Comment (mdebarros) [inline src/stream/producer.js]: This will be lost if the config is set

**Review by mdebarros — CHANGES_REQUESTED** [review]

File: src/stream/producer.js (hunk 1 above)
Comment (rmothilal) [reply src/stream/producer.js:112]: Fixed

File: src/stream/producer.js (hunk 7 above)
Comment (rmothilal) [reply src/stream/producer.js:193]: now used

File: src/stream/producer.js (hunk 8 above)
Comment (rmothilal) [reply src/stream/producer.js]: changed

File: src/stream/producer.js (hunk 9 above)
Comment (rmothilal) [reply src/stream/producer.js]: changed

File: src/stream/producer.js [hunk 10]
```diff
@@ -31,12 +31,397 @@
...
+ * @param {object} config - Key value pairs for the configuration of the Producer with the following:
+ * rdkafkaConf - specific rdkafka configurations [Refer to configuration doc]{@link https://github.com/edenhill/librdkafka/blob/0.11.1.x/CONFIGURATION.md}
+ * topicConf - topic configuration
+ * logger - logger object that supports debug(), info(), verbose(), error() & silly()
+ * @extends EventEmitter
+ * @constructor
+ */
+class Producer extends EventEmitter {
+  constructor (options = {pollIntervalMs: 100}, config = {
+    logger: Logger,
```
Comment (mdebarros) [inline src/stream/producer.js]: options should be in the config!

**Review by mdebarros — APPROVED** [review]

**Pull request merged**